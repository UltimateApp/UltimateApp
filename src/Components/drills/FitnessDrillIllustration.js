import React, { useState, useEffect, useRef } from 'react';
import { Animated, View, StyleSheet, Text, TouchableOpacity, FlatList } from 'react-native';
import { Easing } from 'react-native-reanimated';
import { MaterialCommunityIcons } from '@expo/vector-icons';

import I18n from '../../utils/i18n';
import Animation from '../animation/Animation';
import VimeoVideo from '../shared/VimeoVideo';
import theme from '../../styles/theme.style';
import Drill from '../animation/Drill';

const FitnessDrillIllustration = (props) => {
  const [activeIndex, setActiveIndex] = useState(0);
  const opacityUnchecked = useRef(new Animated.Value(1)).current;
  const opacityChecked = opacityUnchecked.interpolate({ inputRange: [0, 1], outputRange: [1, 0] });
  const currentStep = props.drill.steps[activeIndex];
  const stepsCount = props.drill.steps.length;

  const carouselRef = useRef(null);
  const flatListRef = useRef(null);

  const checkAnimation = () => {
    Animated.sequence([
      Animated.timing(opacityUnchecked, {
        toValue: 0,
        duration: 800,
        easing: Easing.easeOutQuint,
        useNativeDriver: false,
      }),
      Animated.timing(opacityUnchecked, {
        toValue: 1,
        duration: 10,
        useNativeDriver: false,
      }),
    ]).start(() => {
      const newIndex = (activeIndex + 1) % (stepsCount + 1);
      setActiveIndex(newIndex);
      flatListRef.current?.scrollToIndex({ animated: true, index: newIndex, viewPosition: 0.5 });
    });
  };

  // back to 0 when drill changes
  useEffect(() => {
    setActiveIndex(0);
  }, [props.drill]);

  const renderFinish = () => {
    return (
      <View style={styles.containerFinish}>
        <Text style={styles.redoMessage}>{I18n.t('drills.fitnessDrillIllustration.redoMessage')}</Text>
        <TouchableOpacity style={styles.redoButton} onPress={() => setActiveIndex(0)}>
          <MaterialCommunityIcons name="restart" color={theme.COLOR_PRIMARY} size={50} />
        </TouchableOpacity>
      </View>
    );
  };

  const renderStep = ({ index, item }) => {
    const isCurrent = index == activeIndex;

    const doneStyle = index < activeIndex ? styles.stepTitleDone : {};
    const currentStyle = isCurrent ? styles.stepTitleCurrent : {};

    return (
      <TouchableOpacity style={styles.step} onPress={() => setActiveIndex(index)}>
        <View style={styles.subWrapper}>
          <Text style={[styles.stepTitle, doneStyle, currentStyle]}>
            {item.repetition} {item.title}
          </Text>
        </View>
        {isCurrent && (
          <TouchableOpacity style={styles.doneAnimation} onPress={() => checkAnimation()} testID="doneIcon">
            <Animated.View style={{ opacity: opacityUnchecked }}>
              <MaterialCommunityIcons
                style={styles.buttonNext}
                name="check-circle-outline"
                color={theme.COLOR_SECONDARY}
                size={26}
              />
            </Animated.View>
            <Animated.View style={{ opacity: opacityChecked }}>
              <MaterialCommunityIcons
                style={styles.buttonNext}
                name="check-circle"
                color={theme.MAIN_COLOR}
                size={26}
              />
            </Animated.View>
          </TouchableOpacity>
        )}
      </TouchableOpacity>
    );
  };

  const renderVimeo = ({ vimeoId, sounds }) => {
    const isUniqueStep = stepsCount === 1;
    return (
      <>
        <View style={[{ height: 250 }, isUniqueStep && styles.videoAlone]}>
          <VimeoVideo vimeoId={vimeoId} sounds={sounds} shouldPlay />
        </View>
        {!isUniqueStep && (
          <FlatList
            nestedScrollEnabled
            ref={flatListRef}
            data={props.drill.steps}
            keyExtractor={(item) => item.id.toString()}
            renderItem={renderStep}
          />
        )}
      </>
    );
  };

  const renderAnimation = ({ animation, title, instruction }, index) => {
    const isFirstStep = index === 0;
    const isLastStep = index === stepsCount - 1;
    return (
      <>
        <View style={styles.line}>
          <View>
            {!isFirstStep && (
              <TouchableOpacity
                onPress={() => {
                  carouselRef.current.snapToPrev();
                }}
              >
                <MaterialCommunityIcons name="chevron-double-left" color={theme.COLOR_PRIMARY} size={26} />
              </TouchableOpacity>
            )}
          </View>
          <Text style={styles.title}>{title}</Text>
          <View>
            {!isLastStep && (
              <TouchableOpacity
                onPress={() => {
                  carouselRef.current.snapToNext();
                }}
              >
                <MaterialCommunityIcons name="chevron-double-right" color={theme.COLOR_PRIMARY} size={26} />
              </TouchableOpacity>
            )}
          </View>
        </View>
        <View style={styles.animation}>
          <Animation widthRatio={1} heightRatio={1 / 2} animation={new Drill(animation)} />
        </View>
        <Text style={styles.instruction}>{instruction}</Text>
      </>
    );
  };

  return (
    <View style={styles.container}>
      {activeIndex === props.drill.steps.length && renderFinish()}
      {currentStep?.animation && !currentStep?.vimeoId && renderAnimation(currentStep)}
      {currentStep?.vimeoId && renderVimeo(currentStep)}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  containerFinish: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 20,
  },
  redoMessage: {
    fontSize: theme.FONT_SIZE_MEDIUM,
    color: theme.COLOR_PRIMARY,
    fontWeight: 'bold',
    marginVertical: 50,
  },
  redoButton: {
    width: 80,
    height: 80,
  },
  step: {
    flexDirection: 'row',
    padding: 20,
  },
  doneAnimation: {
    flex: 1,
  },
  stepTitle: {
    fontSize: theme.FONT_SIZE_LARGE,
    color: theme.COLOR_PRIMARY,
  },
  stepTitleCurrent: {
    fontWeight: 'bold',
  },
  stepTitleDone: {
    color: theme.COLOR_SECONDARY,
  },
  separator: {
    height: 15,
    borderRightWidth: 1,
    borderRightColor: theme.COLOR_PRIMARY_LIGHT,
  },
  buttonNext: {
    position: 'absolute',
    right: 0,
    marginLeft: 20,
  },
  subWrapper: {
    flexGrow: 0,
    flexShrink: 0,
    alignItems: 'center',
  },
  instruction: {
    marginTop: 20,
    marginBottom: 20,
    marginLeft: 20,
    fontSize: theme.FONT_SIZE_MEDIUM,
    color: theme.COLOR_PRIMARY,
    justifyContent: 'center',
    alignItems: 'center',
  },
  videoAlone: {
    justifyContent: 'center',
    alignItems: 'center',
    flex: 1,
  },
  line: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-around',
    marginTop: 15,
  },
  title: {
    fontSize: theme.FONT_SIZE_LARGE,
    color: theme.COLOR_PRIMARY,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  animation: { flex: 9 },
});

export default FitnessDrillIllustration;
