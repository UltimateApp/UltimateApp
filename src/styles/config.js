export default {
  swipeConfig: {
    velocityThreshold: 0.2,
    directionalOffsetThreshold: 80,
  },
};
