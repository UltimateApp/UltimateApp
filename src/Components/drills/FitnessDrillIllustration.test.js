import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react-native';
import { NavigationContext } from '@react-navigation/native';

import { createDrill } from '../../Fixtures/TestFixtures';
import { DrillTypes } from '../../Fixtures/config';

import FitnessDrillIllustration from './FitnessDrillIllustration';

jest.mock('../shared/VimeoVideo');

describe('<FitnessDrillIllustration />', () => {
  const startFitness = jest.fn();
  const steps = [
    {
      id: 1,
      repetition: 3,
      title: 'First Step',
      vimeoId: '406746924',
    },
    {
      id: 2,
      repetition: 3,
      title: 'Second Step',
      vimeoId: '406747302',
    },
    {
      id: 3,
      repetition: 3,
      title: 'Third Step',
      vimeoId: '406747476',
    },
  ];
  const drill = createDrill({ type: DrillTypes.FITNESS, title: 'Hot Box', steps });

  it('renders correctly with several steps', () => {
    const tree = renderer.create(<FitnessDrillIllustration drill={drill} startFitness={startFitness} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with one step', () => {
    const steps = [
      {
        id: 1,
        repetition: 3,
        title: 'First Step',
        vimeoId: '406746924',
      },
    ];
    const drill = createDrill({ type: DrillTypes.FITNESS, title: 'Hot Box', steps });

    // fake NavigationContext value data
    const navContext = {
      isFocused: () => true,
      // addListener returns an unscubscribe function.
      addListener: jest.fn(() => jest.fn()),
    };

    const tree = renderer
      .create(
        <NavigationContext.Provider value={navContext}>
          <FitnessDrillIllustration drill={drill} startFitness={startFitness} />
        </NavigationContext.Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('triggers startFitness', async () => {
    const { getByTestId } = render(<FitnessDrillIllustration drill={drill} startFitness={startFitness} />);

    fireEvent.press(getByTestId('startFitness'));

    expect(startFitness).toHaveBeenCalled();
  });
});
