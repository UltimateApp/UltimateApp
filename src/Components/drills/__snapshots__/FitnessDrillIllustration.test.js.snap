// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FitnessDrillIllustration /> links to any step and finishes at the end 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "height": 250,
        },
        false,
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "height": 250,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "position": "absolute",
          }
        }
      >
        <ActivityIndicator
          animating={true}
          color="#A6A6A6"
          hidesWhenStopped={true}
          size="large"
        />
        <Text>
          Loading…
        </Text>
      </View>
      <View
        pointerEvents="box-none"
        style={
          Object {
            "height": 250,
            "overflow": "hidden",
            "width": "100%",
          }
        }
      >
        <ViewManagerAdapter_ExpoVideoView
          proxiedProperties={
            Object {
              "onError": [Function],
              "onFullscreenUpdate": [Function],
              "onLoad": [Function],
              "onLoadStart": [Function],
              "onReadyForDisplay": [Function],
              "onStatusUpdate": [Function],
              "resizeMode": "mock",
              "source": undefined,
              "status": Object {},
              "useNativeControls": true,
            }
          }
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "id": 1,
          "repetition": 3,
          "title": "First Step",
          "vimeoId": "406746924",
        },
        Object {
          "id": 2,
          "repetition": 3,
          "title": "Second Step",
          "vimeoId": "406747302",
        },
        Object {
          "id": 3,
          "repetition": 3,
          "title": "Third Step",
          "vimeoId": "406747476",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    nestedScrollEnabled={true}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "opacity": 1,
              "padding": 20,
            }
          }
        >
          <View
            style={
              Object {
                "flexShrink": 1,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#232b2b",
                    "fontSize": 20,
                  },
                  Object {
                    "color": "#A6A6A6",
                  },
                  Object {},
                ]
              }
            >
              3
               
              First Step
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "opacity": 1,
              "padding": 20,
            }
          }
        >
          <View
            style={
              Object {
                "flexShrink": 1,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#232b2b",
                    "fontSize": 20,
                  },
                  Object {
                    "color": "#A6A6A6",
                  },
                  Object {},
                ]
              }
            >
              3
               
              Second Step
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "opacity": 1,
              "padding": 20,
            }
          }
        >
          <View
            style={
              Object {
                "flexShrink": 1,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#232b2b",
                    "fontSize": 20,
                  },
                  Object {},
                  Object {
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              3
               
              Third Step
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "height": "100%",
                "justifyContent": "center",
                "opacity": 1,
                "padding": 10,
              }
            }
            testID="doneIcon"
          >
            <View>
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontSize": 26,
                    },
                    Object {
                      "color": "rgba(166, 166, 166, 1)",
                    },
                    Object {
                      "fontFamily": "material-community",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                󰗠
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`<FitnessDrillIllustration /> links to any step and finishes at the end 2`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#232b2b",
          "fontSize": 16,
          "fontWeight": "bold",
          "marginVertical": 50,
        }
      }
    >
      You have completed the drill!
    </Text>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": 80,
          "opacity": 1,
          "width": 80,
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#232b2b",
              "fontSize": 50,
            },
            undefined,
            Object {
              "fontFamily": "material-community",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        󰜉
      </Text>
    </View>
  </View>
</View>
`;

exports[`<FitnessDrillIllustration /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "height": 250,
        },
        false,
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "height": 250,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "position": "absolute",
          }
        }
      >
        <ActivityIndicator
          animating={true}
          color="#A6A6A6"
          hidesWhenStopped={true}
          size="large"
        />
        <Text>
          Loading…
        </Text>
      </View>
      <View
        pointerEvents="box-none"
        style={
          Object {
            "height": 250,
            "overflow": "hidden",
            "width": "100%",
          }
        }
      >
        <ViewManagerAdapter_ExpoVideoView
          proxiedProperties={
            Object {
              "onError": [Function],
              "onFullscreenUpdate": [Function],
              "onLoad": [Function],
              "onLoadStart": [Function],
              "onReadyForDisplay": [Function],
              "onStatusUpdate": [Function],
              "resizeMode": "mock",
              "source": undefined,
              "status": Object {},
              "useNativeControls": true,
            }
          }
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "id": 1,
          "repetition": 3,
          "title": "First Step",
          "vimeoId": "406746924",
        },
        Object {
          "id": 2,
          "repetition": 3,
          "title": "Second Step",
          "vimeoId": "406747302",
        },
        Object {
          "id": 3,
          "repetition": 3,
          "title": "Third Step",
          "vimeoId": "406747476",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    nestedScrollEnabled={true}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "opacity": 1,
              "padding": 20,
            }
          }
        >
          <View
            style={
              Object {
                "flexShrink": 1,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#232b2b",
                    "fontSize": 20,
                  },
                  Object {},
                  Object {
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              3
               
              First Step
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "height": "100%",
                "justifyContent": "center",
                "opacity": 1,
                "padding": 10,
              }
            }
            testID="doneIcon"
          >
            <View>
              <Text />
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "opacity": 1,
              "padding": 20,
            }
          }
        >
          <View
            style={
              Object {
                "flexShrink": 1,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#232b2b",
                    "fontSize": 20,
                  },
                  Object {},
                  Object {},
                ]
              }
            >
              3
               
              Second Step
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "opacity": 1,
              "padding": 20,
            }
          }
        >
          <View
            style={
              Object {
                "flexShrink": 1,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#232b2b",
                    "fontSize": 20,
                  },
                  Object {},
                  Object {},
                ]
              }
            >
              3
               
              Third Step
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
